tracy-client/src/gpu.rs

```diff
@@ -0,0 +1,163 @@
+use std::sync::atomic::{AtomicU16, AtomicU8, Ordering};
+
+use crate::Client;
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+struct ___tracy_gpu_new_context_data {
+    gpu_time: i64,
+    period: f32,
+    context: u8,
+    flags: u8,
+    type_: u8,
+}
+
+#[repr(u8)]
+///
+pub enum GpuContextType {
+    ///
+    Invalid,
+    ///
+    OpenGL,
+    ///
+    Vulkan,
+    ///
+    OpenCL,
+    ///
+    Direct3D12,
+    ///
+    Direct3D11,
+}
+
+///
+pub struct GpuContext {
+    #[cfg(feature = "enable")]
+    _client: Client,
+    #[cfg(feature = "enable")]
+    context: u8,
+}
+#[cfg(feature = "enable")]
+static GPU_CONTEXT_INDEX: AtomicU8 = AtomicU8::new(0);
+
+///
+pub struct GpuSpan {
+    #[cfg(feature = "enable")]
+    _client: Client,
+    #[cfg(feature = "enable")]
+    context: u8,
+    #[cfg(feature = "enable")]
+    start_query_id: u16,
+    #[cfg(feature = "enable")]
+    end_query_id: Option<u16>,
+}
+#[cfg(feature = "enable")]
+static GPU_SPAN_INDEX: AtomicU16 = AtomicU16::new(0);
+
+impl Client {
+    ///
+    pub fn new_gpu_context(
+        self,
+        name: Option<&str>,
+        ty: GpuContextType,
+        gpu_timestamp: i64,
+        period: f32,
+    ) -> GpuContext {
+        #[cfg(feature = "enable")]
+        unsafe {
+            let context = GPU_CONTEXT_INDEX.fetch_add(1, Ordering::Relaxed);
+
+            sys::___tracy_emit_gpu_new_context(std::mem::transmute(
+                ___tracy_gpu_new_context_data {
+                    gpu_time: gpu_timestamp,
+                    period,
+                    context,
+                    flags: 0,
+                    type_: ty as u8,
+                },
+            ));
+
+            if let Some(name) = name {
+                sys::___tracy_emit_gpu_context_name_serial(sys::___tracy_gpu_context_name_data {
+                    context,
+                    name: name.as_ptr() as *const i8,
+                    len: name.len().min(u16::MAX as usize) as u16,
+                })
+            }
+
+            GpuContext {
+                _client: self,
+                context,
+            }
+        }
+        #[cfg(not(feature = "enable"))]
+        GpuContext {}
+    }
+}
+
+impl GpuContext {
+    ///
+    pub fn span_alloc(&self, name: &str, function: &str, file: &str, line: u32) -> GpuSpan {
+        #[cfg(feature = "enable")]
+        unsafe {
+            let srcloc = sys::___tracy_alloc_srcloc_name(
```

Comment:
We should expose a non-allocating interface as well; see `Client::span` for an example.

Reply:
There is no non-allocating interface for gpu timestamps.

Reply:
To be specific, we should expose an interface which does not call `___tracy_alloc_srcloc_name`. While `___tracy_emit_gpu_zone_begin_alloc_serial` *may* allocate, it uses shared storage internally which is exponentially resized, making it considerably cheaper on average.

Reply:
I need to pull a srcloc from somewhere though right? There's no gpu api that takes source information without allocating.

Reply:
Oh, I see; that's disappointing. Would be nice to open an issue upstream, but not a blocker here, then.